library/core/src/iter/adapters/filter_map.rs

```diff
@@ -39,7 +39,7 @@ fn filter_map_fold<T, B, Acc>(
     }
 }
 
-fn filter_map_try_fold<'a, T, B, Acc, R: Try<Ok = Acc>>(
+fn filter_map_try_fold<'a, T, B, Acc, R: Try<Ok = Acc> + 'a>(
```

Comment:
I think that we could make this change unnecessary with smarter inference. The requirement comes about because we return a closure which captures the lifetime `'a` (via the upvar `f`), so the return type `impl FnMut(Acc, T) -> R + 'a` needs to outlive `'a` rather than `'static`. Therefore, when we return the closure, we end up needing to prove `closure: 'a`, which leads to the requirement that the return type `R` also outlives `'a` (via the new rule introduced by this PR).

Technically, I think we could get away with requiring something weaker than `closure_ty: 'a`. Since the opaque type itself is only known to outlive `'a`, then the 'inputs' to any projection we make using it can only be known to outlive `'a`. Therefore, we can only conclude that the return type `R` outlives `'a`, and not any other region.

However, I don't know how we'd go about expressing that requirement without making some subtle and complicated code even more subtle and complicated.

Reply:
Another thing that might help is that the
```rust
mut fold: impl FnMut(Acc, B) -> R + 'a
```
argument already effectively requires `R: 'a`.

Reply:
Indeed, I am wondering if we can mitigate the impact by leveraging implied bounds.

Reply:
Revisiting this: I don't think that the argument requires `R: 'a`. We could pass in a function pointer for `fold` that looks like:

```rust
fn my_fold<'b>(acc: &'b u8, other: ()) -> &'b u8 { acc }
```

If the `'a` from the original function is `'static`, then we can **not** conclude that `R: 'a` holds.

Reply:
That function should outlive `'static` purely because the `'b` lifetime is constrained by the inputs so it doesn't need to live in the self type and play a role in outlives: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=07b54eba164940719eb62b3b0f151b6b) You're right afaict though that `impl FnMut(T) -> R + 'a` does not prove `R: 'a` since that only requires `/* CLOSURE */: 'a` to hold, you'd also need `T: 'a` to prove that `R: 'a` holds. (maybe its clearer if we write it as `where ANON_T: Fn<(Acc, B), Output = R> + 'a`?)

I think the kind of implied bounds we would want to allow this to compile "as is" is different than the rules that tell us `<T as Trait<U>>::Assoc: 'a if T: 'a, U: 'a`. The reason for proving the return type outlives `'a` when proving `Closure: 'a` is because conceptually we're wanting to avoid the builtin closure impls having unconstrained lifetimes for example:
```rust
impl<'a> Fn<()> for Closure { type Output = &'a () }
``` 
This PR solves that by conceptually having the `Output` type in the `Self` type: 
```rust
impl<'a> Fn<()> for Closure<&'a ()> { type Output = &'a () }
```
Which then means that proving `Closure: 'lifetime` also has to prove `&'a (): 'lifetime`.
(`fn` items currently have a similar workaround where `fn foo<'a>() -> &'a ()` makes `'a` early bound cause the `foo` item to not outlive `'static`.)

If we have a closure with a return type containing a generic param we have to conservatively assume there might be lifetimes in it that are unconstrained, i.e. `|| { T::default() }` called with `T=&'a ()` shouldn't lead to a `Self: 'static` closure. (From what I can tell this seems to be a lot of cases of back incompat changes in this PR?)

So I think the reason we can "prove" that the following is OK:
```rust
fn find<'a, 'b, T, B>(
            f: &'a mut impl FnMut(T) -> Option<B> + 'b,
        ) -> impl FnMut((), T) -> ControlFlow<B> + 'a {
            move |(), x| match f(x) {
                Some(x) => ControlFlow::Break(x),
                None => ControlFlow::CONTINUE,
            }
}
```
is either:
- `B` has to be part of the `Self` type of the return closure because it would have unconstrained lifetimes therefore `impl FnMut((), T) -> ControlFlow<B> + 'a` requires `B: 'a`
    - the argument `f` has the same inputs so if our returned closure had to have `B` in the self type, so would `f`
        - `f` outlives `'b` and `B` is inside `f` and `'b` outlives `'a` therefore `B: 'a`
- `B` does not have to be part of the `Self` type because all of its lifetimes are constrained by input types
    -  pre-this-PR outlives rules only taking into account the `f` upvar which we know outlives `'b` which outlives `'a`

Annoyingly you cant actually know _which_ one of those is true. The important thing is that one of the two must be true, and either case allows you to prove the closure outlives `'a`.